File: src/main/java/com/zhihao/admin/module/system/service/RoleService.java
package com.zhihao.admin.module.system.service;

import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.zhihao.admin.common.api.PageQuery;
import com.zhihao.admin.common.api.PageResult;
import com.zhihao.admin.common.exception.BusinessException;
import com.zhihao.admin.module.system.dto.RoleMenuUpdateRequest;
import com.zhihao.admin.module.system.dto.RoleRequest;
import com.zhihao.admin.module.system.entity.SysRole;
import com.zhihao.admin.module.system.entity.SysRoleMenu;
import com.zhihao.admin.module.system.entity.SysUserRole;
import com.zhihao.admin.module.system.mapper.SysRoleMapper;
import com.zhihao.admin.module.system.mapper.SysRoleMenuMapper;
import com.zhihao.admin.module.system.mapper.SysUserRoleMapper;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class RoleService {

    private final SysRoleMapper roleMapper;
    private final SysRoleMenuMapper roleMenuMapper;
    private final SysUserRoleMapper userRoleMapper;

    public PageResult<SysRole> page(PageQuery query) {
        return PageResult.from(roleMapper.selectPage(Page.of(query.getPageNum(), query.getPageSize()),
                new LambdaQueryWrapper<SysRole>().orderByAsc(SysRole::getSort)));
    }

    @Transactional(rollbackFor = Exception.class)
    public Long create(RoleRequest request) {
        ensureRoleCodeAvailable(request.roleCode(), null);
        SysRole role = new SysRole();
        apply(role, request);
        roleMapper.insert(role);
        return role.getId();
    }

    @Transactional(rollbackFor = Exception.class)
    public void update(Long id, RoleRequest request) {
        SysRole role = requireRole(id);
        ensureRoleCodeAvailable(request.roleCode(), id);
        apply(role, request);
        roleMapper.updateById(role);
    }

    @Transactional(rollbackFor = Exception.class)
    public void delete(Long id) {
        SysRole role = requireRole(id);
        role.setRoleCode(role.getRoleCode() + ":" + role.getId());
        roleMapper.updateById(role);
        roleMenuMapper.delete(new LambdaQueryWrapper<SysRoleMenu>().eq(SysRoleMenu::getRoleId, id));
        userRoleMapper.delete(new LambdaQueryWrapper<SysUserRole>().eq(SysUserRole::getRoleId, id));
        roleMapper.deleteById(id);
    }

    @Transactional(rollbackFor = Exception.class)
    public void updateMenus(Long roleId, RoleMenuUpdateRequest request) {
        requireRole(roleId);
        roleMenuMapper.delete(new LambdaQueryWrapper<SysRoleMenu>().eq(SysRoleMenu::getRoleId, roleId));
        List<SysRoleMenu> roleMenus = request.menuIds().stream().distinct().map(menuId -> {
            SysRoleMenu roleMenu = new SysRoleMenu();
            roleMenu.setRoleId(roleId);
            roleMenu.setMenuId(menuId);
            return roleMenu;
        }).toList();
        if (!roleMenus.isEmpty()) {
            roleMenuMapper.insertBatch(roleMenus);
        }
    }

    public List<Long> getMenuIds(Long roleId) {
        requireRole(roleId);
        return roleMenuMapper.selectList(
                new LambdaQueryWrapper<SysRoleMenu>().eq(SysRoleMenu::getRoleId, roleId)
        ).stream().map(SysRoleMenu::getMenuId).toList();
    }

    private SysRole requireRole(Long id) {
        SysRole role = roleMapper.selectById(id);
        if (role == null) {
            throw new BusinessException("role not found");
        }
        return role;
    }

    private void ensureRoleCodeAvailable(String roleCode, Long ignoreId) {
        SysRole role = roleMapper.selectOne(new LambdaQueryWrapper<SysRole>()
                .eq(SysRole::getRoleCode, roleCode)
                .ne(ignoreId != null, SysRole::getId, ignoreId)
                .last("limit 1"));
        if (role != null) {
            throw new BusinessException("role code already exists");
        }
    }

    private void apply(SysRole role, RoleRequest request) {
        role.setRoleCode(request.roleCode());
        role.setRoleName(request.roleName());
        role.setSort(request.sort());
        role.setStatus(request.status());
        role.setRemark(request.remark());
    }
}

File: server/src/modules/role/role.service.ts
import { formatDateTime } from '../../common/datetime'
import { pageSkip, toPageResult, type PageQuery, type PageResult } from '../../common/page'
import { BusinessException } from '../../common/errors'
import { prisma } from '../../lib/prisma'
import type { RoleMenuUpdateRequest, RoleRequest, SysRoleResponse } from './role.schema'

function toRoleResponse(role: {
  id: bigint
  roleCode: string
  roleName: string
  sort: number | null
  status: number
  remark: string | null
  createTime: Date | null
}): SysRoleResponse {
  return {
    id: Number(role.id),
    roleCode: role.roleCode,
    roleName: role.roleName,
    sort: role.sort ?? 0,
    status: role.status,
    remark: role.remark ?? '',
    createTime: formatDateTime(role.createTime),
  }
}

export async function page(query: PageQuery): Promise<PageResult<SysRoleResponse>> {
  const where = { deleted: 0 }
  const [total, records] = await Promise.all([
    prisma.sysRole.count({ where }),
    prisma.sysRole.findMany({
      where,
      orderBy: { sort: 'asc' },
      skip: pageSkip(query.pageNum, query.pageSize),
      take: query.pageSize,
    }),
  ])
  return toPageResult(total, query.pageNum, query.pageSize, records.map(toRoleResponse))
}

export async function create(request: RoleRequest, operatorId: number | null): Promise<number> {
  await ensureRoleCodeAvailable(request.roleCode, null)
  const now = new Date()
  const role = await prisma.sysRole.create({
    data: {
      roleCode: request.roleCode,
      roleName: request.roleName,
      sort: request.sort ?? 0,
      status: request.status,
      remark: request.remark,
      createTime: now,
      updateTime: now,
      createBy: operatorId !== null ? BigInt(operatorId) : null,
      updateBy: operatorId !== null ? BigInt(operatorId) : null,
      deleted: 0,
    },
  })
  return Number(role.id)
}

export async function update(id: number, request: RoleRequest, operatorId: number | null): Promise<void> {
  await requireRole(id)
  await ensureRoleCodeAvailable(request.roleCode, id)
  await prisma.sysRole.update({
    where: { id: BigInt(id) },
    data: {
      roleCode: request.roleCode,
      roleName: request.roleName,
      sort: request.sort ?? 0,
      status: request.status,
      remark: request.remark,
      updateTime: new Date(),
      updateBy: operatorId !== null ? BigInt(operatorId) : null,
    },
  })
}

export async function remove(id: number): Promise<void> {
  const role = await requireRole(id)
  await prisma.sysRole.update({
    where: { id: BigInt(id) },
    data: { roleCode: `${role.roleCode}:${role.id}`, updateTime: new Date() },
  })
  await prisma.sysRoleMenu.deleteMany({ where: { roleId: BigInt(id) } })
  await prisma.sysUserRole.deleteMany({ where: { roleId: BigInt(id) } })
  await prisma.sysRole.update({
    where: { id: BigInt(id) },
    data: { deleted: 1, updateTime: new Date() },
  })
}

export async function getMenuIds(roleId: number): Promise<number[]> {
  await requireRole(roleId)
  const roleMenus = await prisma.sysRoleMenu.findMany({ where: { roleId: BigInt(roleId) } })
  return roleMenus.map((item) => Number(item.menuId))
}

export async function updateMenus(roleId: number, request: RoleMenuUpdateRequest): Promise<void> {
  await requireRole(roleId)
  await prisma.sysRoleMenu.deleteMany({ where: { roleId: BigInt(roleId) } })
  const menuIds = [...new Set(request.menuIds)]
  if (menuIds.length > 0) {
    await prisma.sysRoleMenu.createMany({
      data: menuIds.map((menuId) => ({
        roleId: BigInt(roleId),
        menuId: BigInt(menuId),
      })),
    })
  }
}

async function requireRole(id: number) {
  const role = await prisma.sysRole.findFirst({
    where: { id: BigInt(id), deleted: 0 },
  })
  if (!role) {
    throw new BusinessException('role not found')
  }
  return role
}

async function ensureRoleCodeAvailable(roleCode: string, ignoreId: number | null) {
  const role = await prisma.sysRole.findFirst({
    where: {
      roleCode,
      deleted: 0,
      ...(ignoreId !== null ? { id: { not: BigInt(ignoreId) } } : {}),
    },
  })
  if (role) {
    throw new BusinessException('role code already exists')
  }
}
